### PR TITLE
Move the repo to newer version of FxCop and Roslyn analyzers (2.3.0-b…

### PR DESCRIPTION
**Pull request by mavasani**
…eta1)

Enable few more rules and fix issues identified these rules.

Also added issue IDs for the disabled rules due to lack of a fixer/fix all experience.

### CODE REVIEW

**Comment by mavasani** [comment]
Tagging @genlu @nguerrera @sharwell @jasonmalinowski. I closed the original PR targeting master https://github.com/dotnet/roslyn/pull/19806 in favor of this one targeting dev15.6

File: src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/StateMachineStates.cs [hunk 1]
```diff
@@ -4,8 +4,8 @@ namespace Microsoft.CodeAnalysis.CSharp
 {
     internal static class StateMachineStates
     {
-        internal readonly static int FinishedStateMachine = -2;
-        internal readonly static int NotStartedStateMachine = -1;
-        internal readonly static int FirstUnusedState = 0;
+        internal const int FinishedStateMachine = -2;
+        internal const int NotStartedStateMachine = -1;
+        internal const int FirstUnusedState = 0;
```
Comment (nguerrera) [inline src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/StateMachineStates.cs]: Is there risk of breaking binary compatibility to things with InternalsVisibleTo access?

Comment (mavasani) [inline src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/StateMachineStates.cs]: I don't think we have ever tracked those for internal API surface, the only concern would be if public APIs are changed.

**Comment by mavasani** [comment]
[windows_debug_vs-integration_prtest](https://ci.dot.net/job/dotnet_roslyn/job/dev15.6/job/windows_debug_vs-integration_prtest/398/) failed with:

```
xunit produced no error output but had exit code 1

Command failed to execute: D:\j\workspace\windows_debug---1522b458\Binaries\Debug\Exes\RunTests\RunTests.exe C:\Users\dotnet-bot\.nuget\packages\xunit.runner.console\2.2.0-beta4-build3444\tools -xml -timeout:50 -procdumppath:c:\SysInternals D:\j\workspace\windows_debug---1522b458\Binaries\Debug\UnitTests\VisualStudioIntegrationTests\Roslyn.VisualStudio.IntegrationTests.dll
10:53:28 System.Management.Automation.RuntimeException: Command failed to execute: D:\j\workspace\windows_debug---1522b458\Binaries\Debug\Exes\RunTests\RunTests.exe C:\Users\dotnet-bot\.nuget\packages\xunit.runner.console\2.2.0-beta4-build3444\tools -xml -timeout:50 -procdumppath:c:\SysInternals D:\j\workspace\windows_debug---1522b458\Binaries\Debug\UnitTests\VisualStudioIntegrationTests\Roslyn.VisualStudio.IntegrationTests.dll
```

@dotnet/roslyn-infrastructure Is this a known issue?

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_vs-integration_prtest please

File: src/VisualStudio/Core/Def/Implementation/Workspace/GlobalUndoServiceFactory.WorkspaceGlobalUndoTransaction.cs [hunk 2]
```diff
@@ -117,11 +117,15 @@ public void Dispose()
                 }
             }
 
+#pragma warning disable CA1821 // Remove empty Finalizers
```
Comment (jinujoseph) [inline src/VisualStudio/Core/Def/Implementation/Workspace/GlobalUndoServiceFactory.WorkspaceGlobalUndoTransaction.cs]: is that not false alert for CA1821 ?

Comment (mavasani) [inline src/VisualStudio/Core/Def/Implementation/Workspace/GlobalUndoServiceFactory.WorkspaceGlobalUndoTransaction.cs]: Yes, tracked with https://github.com/dotnet/roslyn-analyzers/issues/1241

**Comment by mavasani** [comment]
I am merging this in, given I already had sign-offs on the original PR and this is just a rebase to dev15.6 branch. Any further feedback will be addressed with a follow up change.

**Pull request merged**